rfd/0116-label-expressions.md

```diff
@@ -0,0 +1,394 @@
+roles, and 2k unique users.
+
+Benchmark results: TBD
+
+#### Caching
+
+Predicate expressions are parsed and evaluated in two distinct stages.
+
+The first stage takes the raw string expression as input, parses the syntax, and
+returns a closure which is essentially a plain Go function.
+
+The second stage takes closure from the first stage and invokes it with the
+evaluation context (resource labels, user traits) as input to produce a result.
+For label expressions, the result will be a Boolean true/false to indicate
+whether the expression matched the resource.
+
+The first (parse) stage can be easily cached.
+It takes as input the literal string expression which is written in the role,
+and returns a closure that can be invoked repeatedly for each access check with
+the specific inputs for the current user and resource.
+The important part is that is always produces the exact same output for each
+input, and the total number of inputs is relatively small.
+
+The size of the cache is bounded linearly by the number of unique expressions in
+all roles in the cluster.
+These expressions will almost always be hand-written, and I estimate there will
+usually be tens to (possibly) hundreds of unique expressions in any given
+cluster.
+In the rare case that someone writes or generates thousands of label expressions
+that are all actually used, performance can degrade gracefully by using an LRU
+cache.
+
+The cache will be held in-memory of each Teleport process which does access
+checks.
+It will not be shared across nodes, it should be sufficient for each node to
+have to parse each expression once and all future uses can be cached.
+Cache entries will be populated on-demand the first time the are required.
+`github.com/hashicorp/golang-lru/v2` implements a fixed-size thread safe LRU
+cache and is already used for similar purposes (regex caching) within teleport,
+so it will also be used here.
+The default maximum cache size will be 1000, and this can be overridden by an
+environment variable in case it causes problems.
+
+Cache entries never need to be invalidated, updated, or expired, the value for
+each input is valid for the entire lifetime of the process.
+The only thing that could change the parse output is a new version of Teleport
+with a different parsing algorithm, conveniently this will always run in a new
+process with a fresh/empty cache.
+
+Note: an App service will never need to parse or evaluate a
+`db_labels_expression`, it will only cache `app_labels_expression`s, and
+similarly each service will only need to cache the expressions actually relevant
+to that service.
+The only service that will need to parse/cache all expressions is Auth.
+
+### Security
+
+Label expressions, as part of the Role specification, will be editable by any
+users with write/update permissions for Role resources, and readable by any user
+with read permissions for Roles.
+
+This design avoids the possibility of any expression injection attacks.
```

Comment:
I had mentioned these risks earlier in the document also.  I am not sure if we can assume that: 

> No untrusted input will be parsed as an expression

It is possible with cloud for an attacker to create an account and attempt to affect our service through Denial of Service attempts.

I also recognize that the ability write a role is high level of access, but I am not certain that absolves us from needing to consider potential injection attacks from a user who has role write access.

Let me know your thoughts, thank you!

Reply:
How would an attacker creating an account and adding a slow regex to their own tenant benefit them though? Our cloud tenants should be isolated enough such that one slow tenant can not impact the performance of a separate tenant.

Reply:
> Our cloud tenants should be isolated enough such that one slow tenant can not impact the performance of a separate tenant.

This is the root assumption that I am unsure about.  The EKS cluster does utilize shared hardware which is not expected to be fully utilized by every tenant.  How many tenants would I need to create and lock the CPU at 100% until other tenants are impacted?  Even if the number is in the thousands it seems potentially possible an attacker could create that many cloud accounts today.

Just a lot of assumptions around our capacity that I am not certain about.  And even if impacting other tenants is not possible, an attack like this could still be a large financial drain on us.

That said, I think we should address RegEx DoS conditions as part of this issue https://github.com/gravitational/security-findings/issues/27 and not this RFD.  I will resolve this thread and we can continue discussing injection attack risks in the thread above.